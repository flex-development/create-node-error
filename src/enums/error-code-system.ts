/**
 * @file Enums - SystemErrorCode
 * @module errnode/enums/SystemErrorCode
 */

/**
 * System error codes.
 *
 * @see https://github.com/nodejs/node/blob/v19.3.0/deps/uv/include/uv.h#L72-L154
 *
 * @enum {Uppercase<string>}
 */
enum SystemErrorCode {
  E2BIG = 'E2BIG',
  EACCES = 'EACCES',
  EADDRINUSE = 'EADDRINUSE',
  EADDRNOTAVAIL = 'EADDRNOTAVAIL',
  EAFNOSUPPORT = 'EAFNOSUPPORT',
  EAGAIN = 'EAGAIN',
  EAI_ADDRFAMILY = 'EAI_ADDRFAMILY',
  EAI_AGAIN = 'EAI_AGAIN',
  EAI_BADFLAGS = 'EAI_BADFLAGS',
  EAI_BADHINTS = 'EAI_BADHINTS',
  EAI_CANCELED = 'EAI_CANCELED',
  EAI_FAIL = 'EAI_FAIL',
  EAI_FAMILY = 'EAI_FAMILY',
  EAI_MEMORY = 'EAI_MEMORY',
  EAI_NODATA = 'EAI_NODATA',
  EAI_NONAME = 'EAI_NONAME',
  EAI_OVERFLOW = 'EAI_OVERFLOW',
  EAI_PROTOCOL = 'EAI_PROTOCOL',
  EAI_SERVICE = 'EAI_SERVICE',
  EAI_SOCKTYPE = 'EAI_SOCKTYPE',
  EALREADY = 'EALREADY',
  EBADF = 'EBADF',
  EBADMSG = 'EBADMSG',
  EBUSY = 'EBUSY',
  ECANCELED = 'ECANCELED',
  ECHARSET = 'ECHARSET',
  ECHILD = 'ECHILD',
  ECONNABORTED = 'ECONNABORTED',
  ECONNREFUSED = 'ECONNREFUSED',
  ECONNRESET = 'ECONNRESET',
  EDEADLK = 'EDEADLK',
  EDESTADDRREQ = 'EDESTADDRREQ',
  EDOM = 'EDOM',
  EDQUOT = 'EDQUOT',
  EEXIST = 'EEXIST',
  EFAULT = 'EFAULT',
  EFBIG = 'EFBIG',
  EFTYPE = 'EFTYPE',
  EHOSTDOWN = 'EHOSTDOWN',
  EHOSTUNREACH = 'EHOSTUNREACH',
  EIDRM = 'EIDRM',
  EILSEQ = 'EILSEQ',
  EINPROGRESS = 'EINPROGRESS',
  EINTR = 'EINTR',
  EINVAL = 'EINVAL',
  EIO = 'EIO',
  EISCONN = 'EISCONN',
  EISDIR = 'EISDIR',
  ELOOP = 'ELOOP',
  EMFILE = 'EMFILE',
  EMLINK = 'EMLINK',
  EMSGSIZE = 'EMSGSIZE',
  EMULTIHOP = 'EMULTIHOP',
  ENAMETOOLONG = 'ENAMETOOLONG',
  ENETDOWN = 'ENETDOWN',
  ENETRESET = 'ENETRESET',
  ENETUNREACH = 'ENETUNREACH',
  ENFILE = 'ENFILE',
  ENOBUFS = 'ENOBUFS',
  ENODATA = 'ENODATA',
  ENODEV = 'ENODEV',
  ENOENT = 'ENOENT',
  ENOEXEC = 'ENOEXEC',
  ENOLCK = 'ENOLCK',
  ENOLINK = 'ENOLINK',
  ENOMEM = 'ENOMEM',
  ENOMSG = 'ENOMSG',
  ENONET = 'ENONET',
  ENOPROTOOPT = 'ENOPROTOOPT',
  ENOSPC = 'ENOSPC',
  ENOSR = 'ENOSR',
  ENOSTR = 'ENOSTR',
  ENOSYS = 'ENOSYS',
  ENOTCONN = 'ENOTCONN',
  ENOTDIR = 'ENOTDIR',
  ENOTEMPTY = 'ENOTEMPTY',
  ENOTSOCK = 'ENOTSOCK',
  ENOTSUP = 'ENOTSUP',
  ENOTTY = 'ENOTTY',
  ENXIO = 'ENXIO',
  EOF = 'EOF',
  EOPNOTSUPP = 'EOPNOTSUPP',
  EOVERFLOW = 'EOVERFLOW',
  EPERM = 'EPERM',
  EPIPE = 'EPIPE',
  EPROTO = 'EPROTO',
  EPROTONOSUPPORT = 'EPROTONOSUPPORT',
  EPROTOTYPE = 'EPROTOTYPE',
  ERANGE = 'ERANGE',
  EREMOTEIO = 'EREMOTEIO',
  EROFS = 'EROFS',
  ESHUTDOWN = 'ESHUTDOWN',
  ESOCKTNOSUPPORT = 'ESOCKTNOSUPPORT',
  ESPIPE = 'ESPIPE',
  ESRCH = 'ESRCH',
  ESTALE = 'ESTALE',
  ETIME = 'ETIME',
  ETIMEDOUT = 'ETIMEDOUT',
  ETXTBSY = 'ETXTBSY',
  EUNATCH = 'EUNATCH',
  EWOULDBLOCK = 'EWOULDBLOCK',
  EXDEV = 'EXDEV',
  UNKNOWN = 'UNKNOWN'
}

export default SystemErrorCode
