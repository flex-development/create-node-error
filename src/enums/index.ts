/**
 * @file Entry Point - Enums
 * @module errnode/enums
 */

export { default as codes } from './codes'
export { default as syscodes } from './syscodes'
