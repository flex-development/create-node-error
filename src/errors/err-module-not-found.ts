/**
 * @file Errors - ERR_MODULE_NOT_FOUND
 * @module errnode/errors/ERR_MODULE_NOT_FOUND
 * @see https://github.com/nodejs/node/blob/v22.7.0/lib/internal/errors.js#L1581-L1587
 */

import E from '#e'
import { codes } from '#src/enums'
import type { NodeError, NodeErrorConstructor } from '#src/interfaces'

/**
 * `ERR_MODULE_NOT_FOUND` schema.
 *
 * @see {@linkcode NodeError}
 * @see https://nodejs.org/api/errors.html#err_module_not_found
 *
 * @extends {NodeError<codes.ERR_MODULE_NOT_FOUND>}
 */
interface ErrModuleNotFound extends NodeError<codes.ERR_MODULE_NOT_FOUND> {
  /**
   * URL of missing module.
   */
  url?: string | null | undefined
}

/**
 * `ERR_MODULE_NOT_FOUND` message arguments.
 */
type Args = [
  id: string,
  base?: URL | string | null | undefined,
  url?: URL | string | null | undefined
]

/**
 * `ERR_MODULE_NOT_FOUND` constructor.
 *
 * @see {@linkcode Args}
 * @see {@linkcode ErrModuleNotFound}
 * @see {@linkcode NodeErrorConstructor}
 *
 * @extends {NodeErrorConstructor<ErrModuleNotFound,Args>}
 */
interface ErrModuleNotFoundConstructor
  extends NodeErrorConstructor<ErrModuleNotFound, Args> {
  /**
   * Create a new `ERR_MODULE_NOT_FOUND` error.
   *
   * When creating an error for a missing package, `url` should be `null` or
   * `undefined`.
   *
   * @see {@linkcode ErrModuleNotFound}
   *
   * @param {string} id
   *  Id of missing module
   * @param {URL | string | null | undefined} [base]
   *  Parent module id
   * @param {URL | string | null | undefined} [url]
   *  Module URL
   * @return {ErrModuleNotFound}
   */
  new (
    id: string,
    base?: URL | string | null | undefined,
    url?: URL | string | null | undefined
  ): ErrModuleNotFound
}

/**
 * `ERR_MODULE_NOT_FOUND` model.
 *
 * Thrown when a module file cannot be resolved by the ECMAScript modules loader
 * while attempting an `import` operation or when loading a program entry point.
 *
 * @see {@linkcode ErrModuleNotFoundConstructor}
 *
 * @type {ErrModuleNotFoundConstructor}
 *
 * @class
 */
const ERR_MODULE_NOT_FOUND: ErrModuleNotFoundConstructor = E(
  codes.ERR_MODULE_NOT_FOUND,
  Error,
  /**
   * @this {ErrModuleNotFound}
   *
   * @param {string} id
   *  Id of missing module
   * @param {URL | string | null | undefined} [base]
   *  Parent module id
   * @param {URL | string | null | undefined} [url]
   *  Module URL
   * @return {string}
   *  Error message
   */
  function message(
    this: ErrModuleNotFound,
    id: string,
    base: URL | string | null | undefined = null,
    url: URL | string | null | undefined = null
  ): string {
    if (url !== null) this.url = String(url)

    /**
     * Error message.
     *
     * @var {string} message
     */
    let message: string = `Cannot find ${this.url ? 'module' : 'package'}`

    message += ` '${id}'`
    if (base !== null) message += ` imported from ${String(base)}`

    return message
  }
)

export {
  ERR_MODULE_NOT_FOUND as default,
  type ErrModuleNotFound,
  type Args as ErrModuleNotFoundArgs,
  type ErrModuleNotFoundConstructor
}
