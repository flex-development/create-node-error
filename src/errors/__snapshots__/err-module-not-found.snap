// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:errors/ERR_MODULE_NOT_FOUND > should build message from parameters (0) 1`] = `Cannot find module '\${process.cwd()}/missing-file' imported from \${process.cwd()}/src/errors/__tests__/err-module-not-found.spec.ts`;

exports[`unit:errors/ERR_MODULE_NOT_FOUND > should build message from parameters (1) 1`] = `Cannot find package 'missing-package' imported from \${process.cwd()}/src/errors/__tests__/err-module-not-found.spec.ts`;
